group/views.py

```diff
@@ -208,3 +240,80 @@ def edit(request, group_id):
                     return HttpResponseRedirect('/group/detail/%s' % modified_group.id)
         else:
             raise PermissionDenied
+
+##Announce
+@login_required
+def add(request, group_id):
+    group = get_group(group_id)
+
+    if has_group_ownership(request.user, group) or has_group_coownership(request.user, group) or \
+       request.user.has_admin_auth():
+        if request.method == 'POST':
+            form = AnnounceForm(request.POST)
+            if form.is_valid():
+                new_announce = form.save()
+                group.announce.add(new_announce)
+                logger.info('Announce: User %s add Announce %s!' % (request.user.username, new_announce.id))
+                return HttpResponseRedirect(reverse('group:detail', kwargs={'group_id': group_id}))
+            else:
+                return HttpResponseRedirect(reverse('group:detail', kwargs={'group_id': group_id}))
+    else:
+        raise PermissionDenied
+
+@login_required
+def delete_announce(request, announce_id, group_id):
+    group = get_group(group_id)
+
+    if has_group_ownership(request.user, group) or has_group_coownership(request.user, group) or \
```

Comment:
request.user.has_admin_auth() is contained in has_group_ownership, you can eliminate redundant   code.